cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -569,24 +574,25 @@ public void removeApplication(final ApplicationId appId) throws Exception {
    * @throws CannotBeDeletedException : the application cannot be deleted because of running programs
    */
   private void ensureNoRunningPrograms(final ApplicationId appId) throws CannotBeDeletedException {
-    //Check if all are stopped.
-    Iterable<ProgramRuntimeService.RuntimeInfo> runtimeInfos =
-      Iterables.filter(runtimeService.listAll(ProgramType.values()),
-                       new com.google.common.base.Predicate<ProgramRuntimeService.RuntimeInfo>() {
-                         @Override
-                         public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
-                           return runtimeInfo.getProgramId().getParent().equals(appId);
-                         }
-                       });
-
-    Set<String> runningPrograms = new HashSet<>();
-    for (ProgramRuntimeService.RuntimeInfo runtimeInfo : runtimeInfos) {
-      runningPrograms.add(runtimeInfo.getProgramId().getProgram());
-    }
-
-    if (!runningPrograms.isEmpty()) {
+    Map<ProgramRunId, RunRecordMeta> runningPrograms = store.getRuns(ProgramRunStatus.RUNNING, null);
```

Comment:
See if you can avoid scanning across all apps, and just limit to the given app.
Perhaps you can pass in the predicate as the second param here, instead of filtering below. Alternative to this would be to pass in the applicationId to the store.getRuns method.